### PR TITLE
Update 0.5 to Contracts-v2

### PR DESCRIPTION
**Pull request by ben-chain**
This PR updates all the safety checked (`kall`,`kopy`) functions, and all `*Rewrite` functions, to satisfy the `contracts-v2` interfaces and checks.

### CODE REVIEW

File: cmake/EthCompilerSettings.cmake [hunk 1]
```diff
@@ -33,7 +33,7 @@ if (("${CMAKE_CXX_COMPILER_ID}" MATCHES "GNU") OR ("${CMAKE_CXX_COMPILER_ID}" MA
 	# to fix warnings as they arise, so they don't accumulate "to be fixed later".
 	add_compile_options(-Wall)
 	add_compile_options(-Wextra)
-	add_compile_options(-Werror)
+	# add_compile_options(-Werror)
```
Comment (ben-chain) [inline cmake/EthCompilerSettings.cmake:36]: This flag is there to turn all warnings into errors.  For some reason, this always throws on my setup (even vanilla/upstream).  @K-Ho if you haven't run into this we can revert and I'll just keep making the change locally.

File: libsolidity/codegen/CompilerContext.cpp [hunk 2]
```diff
@@ -169,6 +184,14 @@ bool CompilerContext::appendCallback(eth::AssemblyItem const& _i) {
 		ret = true;  // will be set to false again if we don't change the instruction
 		switch (_i.instruction()) {
 			case Instruction::BALANCE:
```
Comment (K-Ho) [inline libsolidity/codegen/CompilerContext.cpp:185]: Same comment as other PR, move `Instruction::SELFBALANCE` above this

File: libsolidity/codegen/CompilerContext.cpp [hunk 3]
```diff
@@ -209,61 +232,100 @@ bool CompilerContext::appendCallback(eth::AssemblyItem const& _i) {
...
-						if eq(success, 0) { revert(0, 0) }
-
+						// technically 0x64 is the minimum needed to add to length, but ABI wants right-padding so we overpad by 0x20.
+						kall(callBytes, add(0x84, length), callBytes, 0x20)
+						// salt is first stack val in ==> first stack val out (address)
 						salt := mload(callBytes)
+						// remove all the stuff we did at callbytes.
+						let newMemSize := msize()
+						for { let ptr := callBytes } lt(ptr, newMemSize) { ptr := add(ptr, 0x20) } {
+							mstore(add(callBytes, ptr), 0x00)
```
Comment (K-Ho) [inline libsolidity/codegen/CompilerContext.cpp]: should this be `mstore(ptr, 0x00)`?

File: libsolidity/codegen/CompilerContext.cpp [hunk 4]
```diff
@@ -209,61 +232,100 @@ bool CompilerContext::appendCallback(eth::AssemblyItem const& _i) {
...
-						if eq(success, 0) { revert(0, 0) }
-
+						// technically 0x44 is the minimum needed to add to length, but ABI wants right-padding so we overpad by 0x20.
+						kall(callBytes, add(0x64, length), callBytes, 0x20)
+						// length is first stack val in ==> first stack val out (address)
 						length := mload(callBytes)
+						// remove all the stuff we did at callbytes.
+						let newMemSize := msize()
+						for { let ptr := callBytes } lt(ptr, newMemSize) { ptr := add(ptr, 0x20) } {
+							mstore(add(callBytes, ptr), 0x00)
```
Comment (K-Ho) [inline libsolidity/codegen/CompilerContext.cpp]: should this be `mstore(ptr, 0x00)`?

**Review by K-Ho — APPROVED** [review]
lgtm!

File: libsolidity/codegen/CompilerContext.cpp (hunk 4 above)
Comment (ben-chain) [reply libsolidity/codegen/CompilerContext.cpp]: yes great catch

File: libsolidity/codegen/CompilerContext.cpp (hunk 3 above)
Comment (ben-chain) [reply libsolidity/codegen/CompilerContext.cpp]: yes great catch

**Pull request merged**